File: ime/src/main/java/com/xuebinduan/ime/MainActivity.java
package com.xuebinduan.ime;

import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.view.WindowInsets;
import android.view.WindowInsetsController;
import android.widget.EditText;
import android.widget.TextView;

/**
 * todo 需要android11，api30及以上才能使用
 * https://medium.com/androiddevelopers/animating-your-keyboard-fb776a8fb66d
 */
public class MainActivity extends AppCompatActivity {

    private EditText editText;
    private TextView textIMEVisibility;
    private TextView textIMESize;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        editText = findViewById(R.id.edit_text);
        textIMEVisibility = findViewById(R.id.text_ime_visibility);
        textIMESize = findViewById(R.id.text_ime_size);

        textIMEVisibility.setOnClickListener(new View.OnClickListener() {
            @RequiresApi(api = Build.VERSION_CODES.R)
            @Override
            public void onClick(View v) {
                //隐藏
                WindowInsetsController windowInsetsController = v.getWindowInsetsController();
                windowInsetsController.hide(WindowInsets.Type.ime());
            }
        });

    }

    @RequiresApi(api = Build.VERSION_CODES.R)
    public void onCheck(View view){
        WindowInsetsCompat insets = ViewCompat.getRootWindowInsets(editText);
        boolean visible= insets.toWindowInsets().isVisible(WindowInsets.Type.ime());
        textIMEVisibility.setText("键盘可见："+(visible?"可见":"不可见"));
        int height = insets.toWindowInsets().getInsets(WindowInsets.Type.ime()).bottom;
        textIMESize.setText("键盘宽高："+height);
    }

}

File: custom_view/OverScroller/src/main/java/com/xuebinduan/overscroller/MainActivity.kt
package com.xuebinduan.overscroller

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle

//https://developer.android.com/training/gestures/scroll
/**
 * OverScroller：
 * startScroll：从指定位置滚动一段指定的距离然后停下，滚动效果与设置的滚动距离、滚动时间、插值器有关，跟离手速度没有关系。一般用于控制 View 滚动到指定的位置
 * fling：从指定位置滑动一段位置然后停下，滚动效果只与离手速度以及滑动边界有关，不能设置滚动距离、滚动时间和插值器。一般用于触摸抬手后继续让 View 滑动一会
 * springBack：从指定位置回弹到指定位置，一般用于实现拖拽后的回弹效果，不能指定回弹时间和插值器
 */
class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.layout_fling)
    }
}